File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/actors/SubframeInterpolableEntity.java
package com.bitfire.uracer.game.actors;

import com.bitfire.uracer.entities.Entity;
import com.bitfire.uracer.entities.EntityRenderState;
import com.bitfire.uracer.game.GameEvents;
import com.bitfire.uracer.game.events.GameRendererEvent;
import com.bitfire.uracer.game.events.PhysicsStepEvent;

public abstract class SubframeInterpolableEntity extends Entity {
    // world-coords
    protected EntityRenderState statePrevious = new EntityRenderState();
    protected EntityRenderState stateCurrent = new EntityRenderState();

    private final PhysicsStepEvent.Listener physicsListener = (source, type, order) -> {
        switch (type) {
            case onBeforeTimestep:
                onBeforePhysicsSubstep();
                break;
            case onAfterTimestep:
                onAfterPhysicsSubstep();
                break;
            case onSubstepCompleted:
                onSubstepCompleted();
                break;
        }
    };

    private final GameRendererEvent.Listener renderListener = (source, type, order) -> {
        if (type == GameRendererEvent.Type.SubframeInterpolate) {
            onSubframeInterpolate(GameEvents.gameRenderer.timeAliasingFactor);
        }
    };

    public SubframeInterpolableEntity() {
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.onBeforeTimestep);
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.onAfterTimestep);
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.onSubstepCompleted);
        GameEvents.gameRenderer.addListener(renderListener, GameRendererEvent.Type.SubframeInterpolate,
                GameRendererEvent.Order.DEFAULT);
    }

    @Override
    public void dispose() {
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.onBeforeTimestep);
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.onAfterTimestep);
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.onSubstepCompleted);
        GameEvents.gameRenderer.removeListener(renderListener, GameRendererEvent.Type.SubframeInterpolate,
                GameRendererEvent.Order.DEFAULT);
    }

    public abstract boolean isVisible();

    public abstract void saveStateTo(EntityRenderState state);

    public abstract boolean isSubframeInterpolated();

    protected void resetState() {
        saveStateTo(stateCurrent);
        statePrevious.set(stateCurrent);
        getStateRender().set(stateCurrent);
        getStateRender().toPixels();
    }

    public void onBeforePhysicsSubstep() {
        saveStateTo(statePrevious);
    }

    public void onAfterPhysicsSubstep() {
        saveStateTo(stateCurrent);
    }

    public void onSubstepCompleted() {

    }

    /**
     * Issued after a tick/physicsStep but before render :P
     */
    public void onSubframeInterpolate(float aliasingFactor) {
        if (isSubframeInterpolated()) {
            if (!EntityRenderState.isEqual(statePrevious, stateCurrent)) {
                getStateRender().set(EntityRenderState.interpolate(statePrevious, stateCurrent, aliasingFactor));
            } else {
                getStateRender().set(stateCurrent);
            }
        } else {
            getStateRender().set(stateCurrent);
        }

        getStateRender().toPixels();
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/actors/SubframeInterpolableEntity.kt
package com.bitfire.uracer.game.actors

import com.bitfire.uracer.entities.Entity
import com.bitfire.uracer.entities.EntityRenderState
import com.bitfire.uracer.game.GameEvents
import com.bitfire.uracer.game.events.GameRendererEvent
import com.bitfire.uracer.game.events.PhysicsStepEvent

abstract class SubframeInterpolableEntity : Entity() {

    // world coordinates
    protected var statePrevious = EntityRenderState()
    protected var stateCurrent = EntityRenderState()

    private val physicsListener = PhysicsStepEvent.Listener { _: Any, type: PhysicsStepEvent.Type, _: PhysicsStepEvent.Order ->
        when (type) {
            PhysicsStepEvent.Type.OnBeforeTimestep -> onBeforePhysicsSubstep()
            PhysicsStepEvent.Type.OnAfterTimestep -> onAfterPhysicsSubstep()
            PhysicsStepEvent.Type.OnSubstepCompleted -> onSubstepCompleted()
        }
    }

    private val renderListener = GameRendererEvent.Listener { _: Any, type: GameRendererEvent.Type, _: GameRendererEvent.Order ->
        if (type == GameRendererEvent.Type.SubframeInterpolate) {
            onSubframeInterpolate(GameEvents.gameRenderer.timeAliasingFactor)
        }
    }

    init {
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.OnBeforeTimestep)
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.OnAfterTimestep)
        GameEvents.physicsStep.addListener(physicsListener, PhysicsStepEvent.Type.OnSubstepCompleted)
        GameEvents.gameRenderer.addListener(renderListener, GameRendererEvent.Type.SubframeInterpolate, GameRendererEvent.Order.DEFAULT)
    }

    override fun dispose() {
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.OnBeforeTimestep)
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.OnAfterTimestep)
        GameEvents.physicsStep.removeListener(physicsListener, PhysicsStepEvent.Type.OnSubstepCompleted)
        GameEvents.gameRenderer.removeListener(renderListener, GameRendererEvent.Type.SubframeInterpolate, GameRendererEvent.Order.DEFAULT)
    }

    abstract val isVisible: Boolean

    abstract fun saveStateTo(state: EntityRenderState)

    abstract val isSubframeInterpolated: Boolean

    protected fun resetState() {
        saveStateTo(stateCurrent)
        statePrevious.set(stateCurrent)
        stateRender.set(stateCurrent)
        stateRender.toPixels()
    }

   open fun onBeforePhysicsSubstep() {
        saveStateTo(statePrevious)
    }

    open fun onAfterPhysicsSubstep() {
        saveStateTo(stateCurrent)
    }

    open fun onSubstepCompleted() {}

    /**
     * Issued after a tick/physicsStep but before render
     */
    fun onSubframeInterpolate(aliasingFactor: Float) {
        if (this.isSubframeInterpolated) {
            if (!EntityRenderState.isEqual(statePrevious, stateCurrent)) {
                stateRender.set(EntityRenderState.interpolate(statePrevious, stateCurrent, aliasingFactor))
            } else {
                stateRender.set(stateCurrent)
            }
        } else {
            stateRender.set(stateCurrent)
        }

        stateRender.toPixels()
    }
}
